raiden-ts/src/transport/epics/messages.ts

```diff
@@ -54,129 +52,182 @@ function getMessageBody(message: string | Signed<Message>): string {
-          return via.label; // via RTC channel, complete immediately
-        });
-      else
-        return matrix$.pipe(
-          mergeMap(async (matrix) => matrix.sendToDevice(type, { [via]: { '*': content } })),
-          mapTo(via), // via toDevice message
-          // complete only when peer is online
-          delayWhen(
-            constant(
-              presence$.pipe(
-                filter(
-                  (presence) => presence?.payload.userId === via && presence.payload.available,
-                ),
-              ),
-            ),
-          ),
-        );
+      return matrix.sendToDevice('m.room.message', payload);
     }),
-    retryWhile(intervalFromConfig(config$), {
+    retryWhile(intervalFromConfig(deps.config$), {
       maxRetries: 3,
       onErrors: networkErrors,
-      log: log.warn,
+      log: deps.log.warn,
+    }),
+    // a higher-order observable is used to be able to have requests serialized/batched,
+    // but return observables for success/failures which will resolve independently when peer comes
+    // online, and get merged in parallel at the end to avoid holding concatMap
+    mergeMap(function* () {
+      const tookMs = Date.now() - start;
+      for (const [peerId, peerQueue] of batch) {
+        for (const action of peerQueue) {
+          // wait independently on each peer's presence when resolving to success
+          yield deps.latest$.pipe(
+            filter(({ presences }) => presences[action.meta.address].payload.available),
+            take(1),
+            mapTo(messageSend.success({ via: peerId, tookMs, retries }, action.meta)),
+          );
+        }
+      }
+    }),
+    catchError(function* (err) {
+      for (const [, peerQueue] of batch) {
+        for (const action of peerQueue) {
+          // if failure, fail all queued responses at once, no need to wait
+          yield of(messageSend.failure(err, action.meta));
+        }
+      }
+    }),
+  );
+}
+
+function sendBatchedToDevice(
+  action$: Observable<messageSend.request>,
+  deps: RaidenEpicDeps,
+): Observable<messageSend.success | messageSend.failure> {
+  const queue: messageSend.request[] = [];
+  return action$.pipe(
+    tap((action) => queue.push(action)),
+    // like concatBuffer, but takes queue control in order to do custom batch selection
+    concatMap(() =>
+      deps.latest$.pipe(
+        first(),
+        // completeWith with some delay gives a minimal time for queue to be flushed
+        mergeMap((latest) =>
+          flushQueueToDevice(queue, latest, deps).pipe(completeWith(action$, 10)),
+        ),
+      ),
+    ),
+    // notice the emitted elements are observables, which can be merged in the output *after*
+    // the concatMap was released in order to "wait" for some condition before success|failure
+    mergeMap((obs$: Observable<messageSend.success | messageSend.failure>) =>
+      obs$.pipe(completeWith(action$, 10)),
+    ),
+  );
+}
+
+function sendNowToRtc(
```

Comment:
`Now` is because `rtc.send` is synchronous, and very fast, and sent right away. This is in contrast with the `Queued` from `toDevice`, which must wait its turn (if there're many messages in queue), and may take a couple of seconds depending on server's congestion and how many messages got queued in a burst.

Reply:
Ah okay. Interesting idea to map to such a naming. But okay. 👍